src/packages/utils/tests/request-coordinator.test.ts

```diff
@@ -0,0 +1,98 @@
+import assert from "assert";
+import { RequestCoordinator } from "../src/utils/request-coordinator";
+
+describe("request-coordinator", () => {
+  const noop = () => undefined;
+  let coordinator: RequestCoordinator;
+
+  beforeEach("instantiate RequestCoordinator", () => {
+    coordinator = new RequestCoordinator(0);
+  });
+
+  describe("stop()", () => {
+    it("should pause processing", () => {
+      coordinator.stop();
+
+      assert(coordinator.paused);
+    });
+
+    it("should not allow processing to be resumed", () => {
+      coordinator.stop();
+
+      assert.throws(
+        () => coordinator.resume(),
+        new Error("Cannot resume processing requests, Ganache is disconnected.")
+      );
+    });
+
+    it("should not allow new requests to be queued", async () => {
+      coordinator.stop();
+
+      await assert.rejects(
+        coordinator.queue(noop, this, []),
+        new Error("Cannot process request, Ganache is disconnected.")
+      );
+    });
+  });
+
+  describe("end()", () => {
+    it("should reject pending requests in the order that they were received", async () => {
+      coordinator.pause();
+
+      let nextRejectionIndex = 0;
+      const pendingAssertions: Promise<any>[] = [];
+
+      for (let taskIndex = 0; taskIndex < 10; taskIndex++) {
+        const task = coordinator.queue(noop, this, []);
+
+        pendingAssertions.push(
+          task.catch(_ => {
```

Comment:
Would this be better written as `.finally(() => {`, since we explicitly check the rejection case below, or is it better to only test the `catch` here?

Reply:
yeah, I think it almost doesn't matter, but finally is maybe more resilient. I can imagine a change to the test meaning that we no longer assert that the task rejects (as we do below), meaning the order check doesn't do anything at all.

Reply:
I reverted this change after thinking about it some more after the test failed :/

I think we can make this a single promise to push into the `pendingAssertions` array, by chaining:

```javascript
const task = coordinator.queue(...).finally(.../*assert order*/)
```

But then it's only checking the rejection order based on the error message, which I don't like.

Alternatively, we could _not_ chain them, but have two assertion arrays, one that all will reject (the `task.finally(...)` ones), and one that all will resolve (`assert.rejects(...)`).

But I think in retrospect I prefer the original approach. Happy to hear alternatives if you have them.

Reply:
:+1: on your original approach!